File: coding-challenges/sources/leetcode/20. Valid Parentheses/20-ValidParentheses.kt
fun main() {
    println(isValid("()") == true)
    println(isValid("()[]{}") == true)
    println(isValid("(]") == false)
}

fun isValid(s: String): Boolean {
    val brackets = mapOf(')' to '(', ']' to '[', '}' to '{')
    val stack = mutableListOf<Char>()
    for (c in s) {
        if (c in brackets.values) {
            stack.add(c)
        } else if (c in brackets) {
            if (stack.isEmpty() || stack.last() != brackets[c]) {
                return false
            }
            stack.removeAt(stack.size - 1)
        }
    }
    
    return stack.isEmpty()
}

File: coding-challenges/sources/leetcode/20. Valid Parentheses/20-ValidParentheses.swift


fileprivate func isValid(_ s: String) -> Bool {
    let brackets: [Character: Character] = [")": "(", "]": "[", "}": "{"]
    var stack = [Character]()
    for c in s {
        if brackets.values.contains(String(c)) {
            stack.append(c)
        } else if let openBracket = brackets[c] {
            if let last = stack.last,
               last != openBracket {
                return false
            }
            stack.removeLast()
        }
    }
    return stack.isEmpty
}

import XCTest

final class ValidParenthesesTests: XCTestCase {
    func test() {
        XCTAssertTrue(isValid("()"))
        XCTAssertTrue(isValid("()[]{}"))
        XCTAssertFalse(isValid("(]"))
    }
}
